internal/tofu/transform_provider.go

```diff
@@ -108,61 +118,78 @@ func (t *ProviderTransformer) Transform(g *Graph) error {
 	type ProviderRequest struct {
-		Addr  addrs.AbsProviderConfig
-		Exact bool // If true, inheritance from parent modules is not attempted
+		Addr        addrs.AbsProviderConfig
+		Exact       ExactProvider // If populated, inheritance from parent modules is not attempted
+		instanceKey addrs.InstanceKey
 	}
 	requested := map[dag.Vertex]map[string]ProviderRequest{}
 	needConfigured := map[string]addrs.AbsProviderConfig{}
 	for _, v := range g.Vertices() {
 		// Does the vertex _directly_ use a provider?
 		if pv, ok := v.(GraphNodeProviderConsumer); ok {
-			providerAddr, exact := pv.ProvidedBy()
-			if providerAddr == nil && exact {
+			providerAddrs, exactProvider := pv.ProvidedBy()
+
+			if providerAddrs == nil && !exactProvider.provider.IsSet() {
 				// no provider is required
 				continue
 			}
 
 			requested[v] = make(map[string]ProviderRequest)
 
-			var absPc addrs.AbsProviderConfig
-
-			switch p := providerAddr.(type) {
-			case addrs.AbsProviderConfig:
-				// ProvidedBy() returns an AbsProviderConfig when the provider
-				// configuration is set in state, so we do not need to verify
-				// the FQN matches.
-				absPc = p
+			if exactProvider.provider.IsSet() {
+				providerAddr := exactProvider.provider
+				log.Printf("[TRACE] ProviderTransformer: %s is provided by %s exactly", dag.VertexName(v), providerAddr)
 
-				if exact {
-					log.Printf("[TRACE] ProviderTransformer: %s is provided by %s exactly", dag.VertexName(v), absPc)
+				requested[v][providerAddr.String()] = ProviderRequest{
+					Addr:  addrs.AbsProviderConfig{},
+					Exact: exactProvider,
+					// If we are providing an exact provider, we won't need to use the instanceKey prop on the request.
+					// That's why we can set addrs.NoKey as the instanceKey.
+					instanceKey: addrs.NoKey,
 				}
 
-			case addrs.LocalProviderConfig:
-				// ProvidedBy() return a LocalProviderConfig when the resource
-				// contains a `provider` attribute
-				absPc.Provider = pv.Provider()
-				modPath := pv.ModulePath()
-				if t.Config == nil {
-					absPc.Module = modPath
-					absPc.Alias = p.Alias
-					break
-				}
+				// Direct references need the provider configured as well as initialized
+				needConfigured[providerAddr.String()] = providerAddr
+			} else {
+				for ik, providerAddr := range providerAddrs {
+					var absPc addrs.AbsProviderConfig
+
+					switch p := providerAddr.(type) {
+					case addrs.AbsProviderConfig:
+						// ProvidedBy() returns an AbsProviderConfig when the provider
+						// configuration is implied from the resource, so we do not need to verify
+						// the FQN matches.
+						absPc = p
+
+					case addrs.LocalProviderConfig:
+						// ProvidedBy() return a LocalProviderConfig when the resource
+						// contains a `provider` attribute
+						absPc.Provider = pv.Provider()
+						modPath := pv.ModulePath()
+						if t.Config == nil {
+							absPc.Module = modPath
+							absPc.Alias = p.Alias
+							break
```

Comment:
just to clarify - this break is not supposed to end the loop, right?

i suggest to put an else statement here to make it more obvious to a reader

Reply:
This is indeed confusing, but the break is on the execution of the specifc switch case